File: android/samples/addbookmarkscreenlet/src/main/java/com/liferay/mobile/screens/bookmark/AddBookmarkScreenlet.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.mobile.screens.bookmark;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import com.liferay.mobile.screens.base.BaseScreenlet;
import com.liferay.mobile.screens.bookmark.interactor.AddBookmarkInteractor;
import com.liferay.mobile.screens.bookmark.interactor.AddBookmarkListener;
import com.liferay.mobile.screens.bookmark.view.AddBookmarkViewModel;

/**
 * @author Javier Gamarra
 */
public class AddBookmarkScreenlet extends BaseScreenlet<AddBookmarkViewModel, AddBookmarkInteractor>
	implements AddBookmarkListener {

	private long folderId;
	private AddBookmarkListener listener;

	public AddBookmarkScreenlet(Context context) {
		super(context);
	}

	public AddBookmarkScreenlet(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	public AddBookmarkScreenlet(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
	}

	public AddBookmarkScreenlet(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
		super(context, attrs, defStyleAttr, defStyleRes);
	}

	public void onAddBookmarkSuccess() {
		getViewModel().showFinishOperation(null);

		if (listener != null) {
			listener.onAddBookmarkSuccess();
		}
	}

	public void onAddBookmarkFailure(Exception e) {
		getViewModel().showFailedOperation(null, e);

		if (listener != null) {
			listener.onAddBookmarkFailure(e);
		}
	}

	public long getFolderId() {
		return folderId;
	}

	public void setFolderId(long folderId) {
		this.folderId = folderId;
	}

	public AddBookmarkListener getListener() {
		return listener;
	}

	public void setListener(AddBookmarkListener listener) {
		this.listener = listener;
	}

	@Override
	protected View createScreenletView(Context context, AttributeSet attributes) {
		TypedArray typedArray =
			context.getTheme().obtainStyledAttributes(attributes, R.styleable.AddBookmarkScreenlet, 0, 0);

		int layoutId = typedArray.getResourceId(R.styleable.AddBookmarkScreenlet_layoutId, 0);

		View view = LayoutInflater.from(context).inflate(layoutId, null);

		String defaultTitle = typedArray.getString(R.styleable.AddBookmarkScreenlet_defaultTitle);

		folderId = castToLong(typedArray.getString(R.styleable.AddBookmarkScreenlet_folderId));

		typedArray.recycle();

		AddBookmarkViewModel viewModel = (AddBookmarkViewModel) view;
		viewModel.setTitle(defaultTitle);

		return view;
	}

	@Override
	protected AddBookmarkInteractor createInteractor(String actionName) {
		return new AddBookmarkInteractor();
	}

	@Override
	protected void onUserAction(String userActionName, AddBookmarkInteractor interactor, Object... args) {
		AddBookmarkViewModel viewModel = getViewModel();
		String url = viewModel.getURL();
		String title = viewModel.getTitle();

		interactor.start(url, title, folderId);
	}
}

File: ios/Samples/Bookmark/AddBookmarkScreenlet/Advanced/AddBookmarkScreenlet.swift
/**
* Copyright (c) 2000-present Liferay, Inc. All rights reserved.
*
* This library is free software; you can redistribute it and/or modify it under
* the terms of the GNU Lesser General Public License as published by the Free
* Software Foundation; either version 2.1 of the License, or (at your option)
* any later version.
*
* This library is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
* FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
* details.
*/
import UIKit
import LiferayScreens

@objc public protocol AddBookmarkScreenletDelegate: BaseScreenletDelegate {

	optional func screenlet(screenlet: AddBookmarkScreenlet,
	                        onBookmarkAdded bookmark: Bookmark)

	optional func screenlet(screenlet: AddBookmarkScreenlet,
	                        onAddBookmarkError error: NSError)
	
}


//Screenlet used for adding bookmarks to a certain folder
public class AddBookmarkScreenlet: BaseScreenlet {


	//MARK: Actions

	static let AddBookmarkAction = "add-bookmark"
	static let GetTitleAction = "get-title"


	//MARK: Inspectables

	@IBInspectable var folderId: Int64 = 0


	///Screenlet ViewModel
	var viewModel: AddBookmarkViewModel {
		return self.screenletView as! AddBookmarkViewModel
	}

	///Screenlet Delegate
	var addBookmarkDelegate: AddBookmarkScreenletDelegate? {
		return self.delegate as? AddBookmarkScreenletDelegate
	}


	//MARK: BaseScreenlet

	override public func createInteractor(name name: String, sender: AnyObject?) -> Interactor? {
		switch name {
		case AddBookmarkScreenlet.AddBookmarkAction:
			return createAddBookmarkInteractor()
		case AddBookmarkScreenlet.GetTitleAction:
			return createGetTitleInteractor()
		default:
			return nil
		}
	}


	//MARK: Private methods

	private func createAddBookmarkInteractor() -> Interactor {
		let interactor = AddBookmarkInteractor(screenlet: self,
		                                       folderId: folderId,
		                                       title: viewModel.title!,
		                                       url: viewModel.URL!)

		//Called when interactor finish succesfully
		interactor.onSuccess = {
            if let bookmark = interactor.resultBookmark {
                self.addBookmarkDelegate?.screenlet?(self, onBookmarkAdded: bookmark)
            }
		}

		//Called when interactor finish with error
		interactor.onFailure = { error in
			self.addBookmarkDelegate?.screenlet?(self, onAddBookmarkError: error)
		}

		return interactor
	}

	private func createGetTitleInteractor() -> Interactor {
        let interactor = GetWebTitleInteractor(screenlet: self, url: viewModel.URL!)

		//Called when interactor finish succesfully
		interactor.onSuccess = {
			let title = interactor.resultTitle
			self.viewModel.title = title
		}

		//Called when interactor finish with error
		interactor.onFailure = { _ in
			print("An error occurred retrieving the title")
		}

		return interactor
	}
}
